File: adk-go/src/interfaces/cli_runner.go
package interfaces

import (
    "context"
    "fmt"
    "strings"
    "time"

    "adk-benchmark/src/application/factories"
    "adk-benchmark/src/domain"
    "adk-benchmark/src/infrastructure"

    "google.golang.org/adk/agent"
    "google.golang.org/adk/runner"
    "google.golang.org/adk/session"
    "google.golang.org/genai"
)

func RunScenario(ctx context.Context, modelName string, debug bool) {
    fmt.Printf("[URUCHAMIANIE] Model: %s | Framework: ADK-Go\n", modelName)

    metrics := infrastructure.NewMetricsService(modelName)
    metrics.SampleRAM()

    ag, err := factories.BuildAgent(ctx, modelName)
    if err != nil {
        fmt.Printf("Błąd budowania agenta: %v\n", err)
        return
    }

    sessionService := session.InMemoryService()
    r, err := runner.New(runner.Config{
        AppName:        "invoice-benchmark",
        Agent:          ag,
        SessionService: sessionService,
    })
    if err != nil {
        fmt.Printf("Błąd tworzenia runnera: %v\n", err)
        return
    }

    respCreate, err := sessionService.Create(ctx, &session.CreateRequest{
        AppName: "invoice-benchmark",
        UserID:  "bench-user",
    })
    if err != nil {
        fmt.Printf("Błąd tworzenia sesji: %v\n", err)
        return
    }
    sessionID := respCreate.Session.ID()

    for i, userText := range domain.StandardScenario {
        fmt.Printf("\n[Wiadomość %d/%d]\n", i+1, len(domain.StandardScenario))

        preview := userText
        if len(preview) > 80 {
            preview = preview[:80] + "..."
        }
        fmt.Printf("Użytkownik: %s\n", preview)

        metrics.LogTurn("user", userText, nil)
        turnStart := time.Now()
        metrics.SampleRAM()

        agentResponse := ""
        toolCallsThisTurn := []string{}

        if debug {
            fmt.Println("  \033[90m[DEBUG] Czekam na odpowiedź agenta...\033[0m")
        }

        var content *genai.Content
        contentList := genai.Text(userText)
        if len(contentList) > 0 {
            content = contentList[0]
        }
        
        events := r.Run(ctx, "bench-user", sessionID, content, agent.RunConfig{})

        tokensIn, tokensOut := 0, 0
        firstResponseRecorded := false

        func() {
            defer func() {
                if p := recover(); p != nil {
                    errMsg := fmt.Sprintf("%v", p)
                    if debug {
                        fmt.Printf("\n  \033[91m[DEBUG ERROR] %s\033[0m\n", errMsg)
                    }
                    if strings.Contains(errMsg, "not found") && strings.Contains(errMsg, "Tool") {
                        agentResponse = "[BŁĄD: model wywołał nieistniejące narzędzie]"
                        toolCallsThisTurn = append(toolCallsThisTurn, "HALLUCINATED_TOOL")
                        metrics.RecordLLMCall(0, 0)
                        if !firstResponseRecorded {
                            metrics.RecordFirstResponse()
                            firstResponseRecorded = true
                        }
                    } else {
                        panic(p)
                    }
                }
            }()

            for event := range events {
			if event == nil {
				continue
			}

			if !firstResponseRecorded {
				metrics.RecordFirstResponse()
				firstResponseRecorded = true
			}

			if event.Content != nil {
				for _, part := range event.Content.Parts {
					if part == nil {
						continue
					}
					
					if part.Text != "" {
						if debug {
							fmt.Printf("\033[90m%s\033[0m", part.Text)
						}
						agentResponse += part.Text
					}
					
					if part.FunctionCall != nil {
						toolCallsThisTurn = append(toolCallsThisTurn, part.FunctionCall.Name)
						if debug {
							fmt.Printf("\n  \033[93m[DEBUG] Agent wywołuje narzędzie: %s\033[0m\n", part.FunctionCall.Name)
						}
					}
				}
			}

			if event.UsageMetadata != nil {
				tokensIn = int(event.UsageMetadata.PromptTokenCount)
				tokensOut = int(event.UsageMetadata.CandidatesTokenCount)
			}
		}
		}()

        if debug {
            fmt.Println()
        }

        metrics.RecordLLMCall(tokensIn, tokensOut)
        metrics.SampleRAM()  

        turnTime := time.Since(turnStart).Seconds()

		if strings.Contains(agentResponse, "\"invoice\"") && (strings.Contains(agentResponse, "\"totals\"") || strings.Contains(agentResponse, "\"financials\"")) {
			extracted := infrastructure.ExtractInvoiceJSON(agentResponse)
			if extracted != nil {
				metrics.FinalInvoiceJSON = extracted
			}
		}
        
        if !debug {
            respPreview := agentResponse
            if len(respPreview) > 120 {
                respPreview = respPreview[:120] + "..."
            }
            fmt.Printf("Agent: %s\n", respPreview)
            fmt.Printf("[%.3fs | użyte narzędzia: %v]\n", turnTime, toolCallsThisTurn)
        } else {
            fmt.Printf("\n  \033[92m[ZAKOŃCZONO TURĘ w %.3fs. Narzędzia: %v]\033[0m\n", turnTime, toolCallsThisTurn)
        }

        metrics.LogTurn("agent", agentResponse, map[string]any{
            "turn_time_s": turnTime,
            "tool_calls":  toolCallsThisTurn,
        })
    }

    metrics.Finish()
    metrics.Quality = infrastructure.EvaluateQuality(metrics.ConversationLog)
    savedPath := metrics.Save()

    dict := metrics.ToDict()
    score, ok := metrics.Quality["auto_score_0_5"].(float64)
    if !ok {
        score = 0.0
    }
    fmt.Printf("\n[PODSUMOWANIE %s] Czas: %.3fs | Ocena: %.2f/5\n", modelName, dict["total_time_s"], score)
    fmt.Printf("Zapisano w: %s\n", savedPath)
}

func ExecuteAll(debug bool) {
    ctx := context.Background()
    for _, model := range domain.ModelsToTest {
        RunScenario(ctx, model, debug)
    }
}

File: adk-typescript/src/interfaces/cli_runner.ts
import { Runner, InMemorySessionService } from '@google/adk';
import { buildAgent } from '../application/factories/agent_factory';
import { StandardScenario } from '../domain/config';
import { ModelsToTest } from '../domain/constants';
import { MetricsService } from '../infrastructure/metrics_service';
import { extractInvoiceJson, evaluateQuality } from '../infrastructure/evaluation';

export async function runScenario(modelName: string, debug: boolean) {
    console.log(`[URUCHAMIANIE] Model: ${modelName} | Framework: ADK-TypeScript`);

    const metrics = new MetricsService(modelName);
    metrics.sampleRam();

    let agent;
    try {
        agent = buildAgent(modelName);
    } catch (err) {
        console.error(`Błąd budowania agenta: ${err}`);
        return;
    }

    const sessionService = new InMemorySessionService();
    const runner = new Runner({
        appName: "invoice-benchmark",
        agent: agent,
        sessionService: sessionService,
    });

    const session = await sessionService.createSession({ userId: "bench-user", appName: "invoice-benchmark" } as any);
    const sessionId = session.id;

    for (let i = 0; i < StandardScenario.length; i++) {
        const userText = StandardScenario[i];
        console.log(`\n[Wiadomość ${i + 1}/${StandardScenario.length}]`);
        
        let preview = userText;
        if (preview.length > 80) preview = preview.substring(0, 80) + "...";
        console.log(`Użytkownik: ${preview}`);

        metrics.logTurn("user", userText, null);
        const turnStart = Date.now();
        metrics.sampleRam();

        let agentResponse = "";
        let toolCallsThisTurn: string[] = [];

        if (debug) {
            console.log("  \x1b[90m[DEBUG] Czekam na odpowiedź agenta...\x1b[0m");
        }

        let tokensIn = 0;
        let tokensOut = 0;
        let firstResponseRecorded = false;

        try {
            const stream = await runner.runAsync({
                userId: "bench-user", 
                sessionId: sessionId, 
                newMessage: { role: "user", parts: [{ text: userText }] }
            });

            for await (const evt of stream) {
                if (!firstResponseRecorded) {
                    metrics.recordFirstResponse();
                    firstResponseRecorded = true;
                }
                const event = evt as any;

                if (event.content?.parts) {
                    for (const part of event.content.parts) {
                        if (part.text) {
                            if (debug) {
                                process.stdout.write(`\x1b[90m${part.text}\x1b[0m`);
                            }
                            agentResponse += part.text;
                        }
                        if (part.functionCall) {
                            toolCallsThisTurn.push(part.functionCall.name);
                            if (debug) {
                                console.log(`\n  \x1b[93m[DEBUG] Agent wywołuje narzędzie: ${part.functionCall.name}\x1b[0m`);
                            }
                        }
                    }
                }

                if (event.usageMetadata) {
                    tokensIn = event.usageMetadata.promptTokens || 0;
                    tokensOut = event.usageMetadata.candidatesTokenCount || event.usageMetadata.completionTokens || 0;
                }
            }
        } catch (err: any) {
            if (debug) {
                console.log(`\n  \x1b[91m[DEBUG ERROR] ${err.message || err}\x1b[0m`);
            }
            if (String(err).includes("Tool") || String(err).includes("not found")) {
                agentResponse = "[BŁĄD: model wywołał nieistniejące narzędzie]";
                toolCallsThisTurn.push("HALLUCINATED_TOOL");
                metrics.recordLlmCall(0, 0);
                if (!firstResponseRecorded) {
                    metrics.recordFirstResponse();
                    firstResponseRecorded = true;
                }
            } else {
                throw err;
            }
        }

        if (debug) {
            console.log("");
        }

        metrics.recordLlmCall(tokensIn, tokensOut);
        metrics.sampleRam();

        const turnTime = (Date.now() - turnStart) / 1000.0;

        if (agentResponse.includes('"invoice"') && (agentResponse.includes('"totals"') || agentResponse.includes('"financials"'))) {
            const extracted = extractInvoiceJson(agentResponse);
            if (extracted) {
                metrics.finalInvoiceJson = extracted;
            }
        }

        if (!debug) {
            let respPreview = agentResponse;
            if (respPreview.length > 120) respPreview = respPreview.substring(0, 120) + "...";
            console.log(`Agent: ${respPreview}`);
            console.log(`[${turnTime.toFixed(3)}s | użyte narzędzia: [${toolCallsThisTurn.join(", ")}]]`);
        } else {
            console.log(`\n  \x1b[92m[ZAKOŃCZONO TURĘ w ${turnTime.toFixed(3)}s. Narzędzia: [${toolCallsThisTurn.join(", ")}]]\x1b[0m`);
        }

        metrics.logTurn("agent", agentResponse, {
            turn_time_s: turnTime,
            tool_calls: toolCallsThisTurn
        });
    }

    metrics.finish();
    metrics.quality = evaluateQuality(metrics.conversationLog);
    const savedPath = metrics.save();

    const dict = metrics.toDict();
    let score = dict.quality && dict.quality.auto_score_0_5 ? dict.quality.auto_score_0_5 : 0.0;
    
    console.log(`\n[PODSUMOWANIE ${modelName}] Czas: ${dict.total_time_s.toFixed(3)}s | Ocena: ${score.toFixed(2)}/5`);
    console.log(`Zapisano w: ${savedPath}`);
}

export async function executeAll(debug: boolean) {
    for (const model of ModelsToTest) {
        await runScenario(model, debug);
    }
}
